src/rdkafka_buf.h

```diff
@@ -951,11 +967,60 @@ static RD_INLINE void rd_kafka_buf_update_u32 (rd_kafka_buf_t *rkbuf,
 }
 
 
+/**
+ * @brief Write array count field to buffer (i32) for later update with
+ *        rd_kafka_buf_update_arraycnt().
+ */
+#define rd_kafka_buf_write_arraycnt_pos(rkbuf) rd_kafka_buf_write_i32(rkbuf, 0)
+
+
+/**
+ * @brief Write the final array count to the position returned from
+ *        rd_kafka_buf_write_arraycnt_pos().
+ *
+ * Update int32_t in buffer at offset 'of' but serialize it as
+ * compact uvarint (that must not exceed 4 bytes storage)
+ * if the \p rkbuf is marked as FLEXVER, else just update it as
+ * as a standard update_i32().
+ *
+ * @remark For flexibleVersions this will shrink the buffer and move data
+ *         and may thus be costly.
```

Comment:
ouch.

it looks like it's feasible to compute the size of these ahead of time (where required, currently anyway), and that would be more efficient.

or, maybe a better strategy to initially allocate a smaller number of bytes for arraycnt (whatever is most commonly expected, possibly have that configurable per usage), and insert bytes if the estimate isn't right.

but optimization. if it works i don't care, task for later PR.

Reply:
memmoves are pretty fast since everything is in cache.
Pre-calculating a multi-tiered list of topics and partition sizes is more work.